### PR TITLE
Implement Convolution with cuDNN

### PR DESCRIPTION
**Pull request by targetsm**

### CODE REVIEW

File: dace/frontend/tensorflow/tensorflow.py [hunk 1]
```diff
@@ -497,7 +503,19 @@ def compile(self,
             ############################
             # Compile the SDFG
             if gpu:
-                #    self.graph.apply_gpu_transformations()
+
+                from dace.transformation.interstate import GPUTransformSDFG
+                self.graph.apply_transformations([GPUTransformSDFG], apply_once=True)
+                
+                #Do not transform cuDNN tasklets
```
Comment (tbennun) [inline dace/frontend/tensorflow/tensorflow.py]: only if cudnn is enabled

File: dace/frontend/tensorflow/tensorflow.py [hunk 2]
```diff
@@ -367,6 +367,12 @@ def compile(self,
         """
         from dace.config import Config
 
+        path = ' -I/usr/local/cuda-10.1/include'
```
Comment (tbennun) [inline dace/frontend/tensorflow/tensorflow.py]: fixed paths

File: dace/frontend/tensorflow/tensorflow.py [hunk 3]
```diff
@@ -1981,6 +2001,187 @@ def visit_Conv2D(self, node):
...
+                                             yDesc, y));
+                '''.format(N=N, C=C, H=H, W=W, K=K, R=R, S=S, padh=padh, padw=padw,
+                           vstr=strides[1], hstr=strides[2], dilh=dilations[1], dilw=dilations[2]),
+                language=dace.Language.CPP,
+                location="gpu"
+            )
+            code_global = ''' 
+                #include <cudnn.h>
+                #include <iostream>
+                cudnnHandle_t cudnn_handle;
```
Comment (tbennun) [inline dace/frontend/tensorflow/tensorflow.py]: move to graph.global_code

File: dace/frontend/tensorflow/tensorflow.py [hunk 4]
```diff
@@ -1981,6 +2001,187 @@ def visit_Conv2D(self, node):
...
+            '''         
+            if not self.cudnn:
```
Comment (tbennun) [inline dace/frontend/tensorflow/tensorflow.py]: ???

File: dace/frontend/tensorflow/tensorflow.py [hunk 5]
```diff
@@ -1981,6 +2001,187 @@ def visit_Conv2D(self, node):
...
+                     checkCUDNN(cudnnGetConvolutionForwardWorkspaceSize(cudnn_handle,
+                                                             xDesc,
+                                                             fDesc,
+                                                             convDesc,
+                                                             yDesc,
+                                                             algo,
+                                                             &workSpaceSizeInBytes));
+
+                     void* workSpace{{nullptr}};
+                     cudaMalloc(&workSpace, workSpaceSizeInBytes);
```
Comment (tbennun) [inline dace/frontend/tensorflow/tensorflow.py]: does not belong in `code`

File: dace/frontend/tensorflow/tensorflow.py [hunk 6]
```diff
@@ -497,7 +503,19 @@ def compile(self,
...
+                
+                #Do not transform cuDNN tasklets
+                for state in self.graph.nodes():
+                    for node in state.nodes():
+                        if node.label in ['Conv2D_gmap', 
+                                          'Conv2DBackpropInput_gmap', 
+                                          'Conv2DBackpropFilter_gmap']:
+                            node.map.schedule = dace.ScheduleType.Sequential
+                
+                #self.graph.apply_gpu_transformations()
```
Comment (tbennun) [inline dace/frontend/tensorflow/tensorflow.py]: unnecessary, change xform

File: dace/transformation/interstate/gpu_transform_sdfg.py [hunk 7]
```diff
@@ -246,6 +246,9 @@ def apply(self, sdfg: sd.SDFG):
 
         for state, gcodes in zip(sdfg.nodes(), global_code_nodes):
             for gcode in gcodes:
+                # Exception for cuDNN
+                if gcode.label in ['Conv2D','Conv2DBackpropInput','Conv2DBackpropFilter']:
+                    continue
```
[the remark this answers is not in the document]
Comment (targetsm) [reply dace/transformation/interstate/gpu_transform_sdfg.py]: I went over it, should be ok now

File: samples/sdfg_api/cudnn_tasklet.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    code='''
+    // Set the current stream to match DaCe (for correct synchronization)
+    cudnnSetStream(handle, __dace_current_stream);          
+    #define checkCUDNN(expression)                           \\
+    {{                                                        \\
+      cudnnStatus_t status = (expression);                   \\
+      if (status != CUDNN_STATUS_SUCCESS) {{                  \\
+        printf(\"%s \\n \\n \", cudnnGetErrorString(status));                \\
+      }}                                                      \\
+    }}
```
[the remark this answers is not in the document]
Comment (targetsm) [reply samples/sdfg_api/cudnn_tasklet.py]: Replaced it

File: dace/codegen/targets/cuda.py [hunk 9]
```diff
@@ -152,8 +152,6 @@ def get_generated_codeobjects(self):
 #include <cuda_runtime.h>
 #include <dace/dace.h>
 
-{file_header}
-
```
Comment (tbennun) [inline dace/codegen/targets/cuda.py]: Why?

File: tests/cuda_test.sh [hunk 10]
```diff
@@ -184,6 +184,8 @@ runall() {
     runopt samples/simple/axpy.py $1 'GPUTransformSDFG$0'
     
     runtestargs instrumentation_test.py gpu
+	
+	runtestargs tensorflow/conv_test.py gpu
```
Comment (tbennun) [inline tests/cuda_test.sh]:
```suggestion
    runtestargs tensorflow/conv_test.py gpu
```

File: samples/sdfg_api/cudnn_tasklet.py [hunk 11]
```diff
@@ -46,15 +46,43 @@
...
+    code_global='''
+    #include <cudnn.h>
+    cudnnHandle_t handle;
+    cudnnTensorDescriptor_t xDesc;
+    cudnnTensorDescriptor_t yDesc;
+    cudnnFilterDescriptor_t fDesc;
+    cudnnConvolutionDescriptor_t convDesc;
+    cudnnConvolutionFwdAlgo_t algo;
+    size_t workSpaceSizeInBytes = 0;
+    void* workSpace{nullptr};
```
Comment (tbennun) [inline samples/sdfg_api/cudnn_tasklet.py]: you could keep these inside `code` for the sample.

File: dace/codegen/targets/cuda.py (hunk 9 above)
Comment (targetsm) [reply dace/codegen/targets/cuda.py]: It doesn't work even for the tests with this line...

File: samples/sdfg_api/cudnn_tasklet.py (hunk 11 above)
Comment (targetsm) [reply samples/sdfg_api/cudnn_tasklet.py]: Well it should go to init then. Or all of init also to code?

Comment (tbennun) [reply samples/sdfg_api/cudnn_tasklet.py]: all of init can go into the code for this sample, similarly to the CUBLAS sample

**Pull request closed without merging**